File: MISTRAL_NKO_VOICE_TRANSCRIT_18_07_2026/android-app/app/src/main/java/com/nko/voice/api/ApiClient.kt
package com.nko.voice.api

import com.nko.voice.BuildConfig
import java.io.File
import java.io.IOException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

data class TranscriptionResult(
    val latinText: String,
    val nkoText: String,
    val engine: String,
    val durationMs: Long,
)

class ApiClient(
    private val tokens: TokenStore? = null,
    private val baseUrl: String = BuildConfig.API_BASE_URL,
) {
    private val http = OkHttpClient()

    fun login(email: String, password: String) {
        val body = JSONObject().put("email", email).put("password", password)
        val request = Request.Builder()
            .url("$baseUrl/api/v1/auth/login")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw IOException("Login failed: ${response.code}")
            val json = JSONObject(response.body!!.string())
            tokens?.accessToken = json.getString("access_token")
            tokens?.refreshToken = json.getString("refresh_token")
        }
    }

    fun transcribe(
        audio: File,
        language: String = "bam",
        engine: String? = null,
        storeHistory: Boolean = false,
    ): TranscriptionResult {
        val multipart = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart(
                "audio", audio.name,
                audio.asRequestBody("audio/mp4".toMediaType()),
            )
            .addFormDataPart("language", language)
            .addFormDataPart("store_history", storeHistory.toString())
            .apply { engine?.let { addFormDataPart("asr_engine", it) } }
            .build()
        val builder = Request.Builder()
            .url("$baseUrl/api/v1/transcriptions/upload")
            .post(multipart)
        tokens?.accessToken?.let { builder.header("Authorization", "Bearer $it") }
        http.newCall(builder.build()).execute().use { response ->
            if (!response.isSuccessful) throw IOException("Transcription failed: ${response.code}")
            return parseTranscription(response.body!!.string())
        }
    }

    companion object {
        fun parseTranscription(payload: String): TranscriptionResult {
            val json = JSONObject(payload)
            return TranscriptionResult(
                latinText = json.getString("latin_text"),
                nkoText = json.getString("nko_text"),
                engine = json.getString("asr_engine"),
                durationMs = json.optLong("duration_ms"),
            )
        }
    }
}

File: MISTRAL_NKO_VOICE_TRANSCRIT_18_07_2026/ios-app/NkoTranscribe/APIClient.swift
import Foundation

struct TranscriptionResult: Decodable, Equatable {
    let latinText: String
    let nkoText: String
    let asrEngine: String
    let durationMs: Int?

    enum CodingKeys: String, CodingKey {
        case latinText = "latin_text"
        case nkoText = "nko_text"
        case asrEngine = "asr_engine"
        case durationMs = "duration_ms"
    }
}

struct TokenPair: Decodable {
    let accessToken: String
    let refreshToken: String

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
    }
}

enum APIError: Error {
    case http(Int)
    case invalidResponse
}

final class APIClient {

    private let session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = Config.requestTimeout
        return URLSession(configuration: configuration)
    }()

    func login(email: String, password: String) async throws {
        var request = URLRequest(url: Config.baseURL.appending(path: "/api/v1/auth/login"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(["email": email, "password": password])
        let (data, response) = try await session.data(for: request)
        try Self.checkStatus(response)
        let tokens = try JSONDecoder().decode(TokenPair.self, from: data)
        KeychainStore.set(tokens.accessToken, for: "access_token")
        KeychainStore.set(tokens.refreshToken, for: "refresh_token")
    }

    func transcribe(fileURL: URL, language: String = Config.defaultLanguage)
        async throws -> TranscriptionResult
    {
        let boundary = "nko-\(UUID().uuidString)"
        var request = URLRequest(
            url: Config.baseURL.appending(path: "/api/v1/transcriptions/upload"))
        request.httpMethod = "POST"
        request.setValue(
            "multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        if let token = KeychainStore.get("access_token") {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let audioData = try Data(contentsOf: fileURL)
        request.httpBody = Self.multipartBody(
            boundary: boundary,
            audio: audioData,
            filename: fileURL.lastPathComponent,
            fields: ["language": language]
        )
        let (data, response) = try await session.data(for: request)
        try Self.checkStatus(response)
        return try JSONDecoder().decode(TranscriptionResult.self, from: data)
    }

    static func multipartBody(
        boundary: String, audio: Data, filename: String, fields: [String: String]
    ) -> Data {
        var body = Data()
        for (name, value) in fields {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append(
                "Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n\(value)\r\n"
                    .data(using: .utf8)!)
        }
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append(
            "Content-Disposition: form-data; name=\"audio\"; filename=\"\(filename)\"\r\n"
                .data(using: .utf8)!)
        body.append("Content-Type: audio/mp4\r\n\r\n".data(using: .utf8)!)
        body.append(audio)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        return body
    }

    private static func checkStatus(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else { throw APIError.http(http.statusCode) }
    }
}
